src/main/java/com/android/volley/AsyncRequestQueue.java

```diff
@@ -119,34 +131,37 @@ public void start() {
         mNetwork.setNonBlockingExecutor(mNonBlockingExecutor);
         mNetwork.setNonBlockingScheduledExecutor(mNonBlockingScheduledExecutor);
 
-        mNonBlockingExecutor.execute(
-                new Runnable() {
-                    @Override
-                    public void run() {
-                        // This is intentionally blocking, because we don't want to process any
-                        // requests until the cache is initialized.
-                        if (mAsyncCache != null) {
-                            final CountDownLatch latch = new CountDownLatch(1);
+        // Kick off cache initialization, which must complete before any requests can be processed.
+        if (mAsyncCache != null) {
```

Comment:
It's out of scope here, but it would be a bit easier to follow if the notion of initializing cache was abstracted, so `start()` doesn't have to deal directly with checking which cache is in use and just try to initialize "a" cache.

Since it requires a refactor of the class, I don't think we should do it now. But something to keep in mind as a future clean up, when we do further modifications to the async code (like making it thread safe?)

Reply:
It could make sense to fork off some helper classes from AsyncRequestQueue for the cache and network layers, similar to how CacheDispatcher / NetworkDispatcher end up abstracting those details away from RequestQueue. I'm hesitant to push it all the way down to the cache layer unless we can find a way to avoid having to duplicate logic in each implementation. (We're somewhat hampered by Cache being an interface rather than an abstract class, which prevents us from being able to expose equivalent initialization methods in both.)